### PR TITLE
feat: `hide-cast-button` patch

### PR DESCRIPTION
**Pull request by ghost**
Note: Signature is not working yet

### CODE REVIEW

File: src/main/kotlin/app/revanced/patches/youtube/layout/castbutton/annotations/CastPatchCompatibility.kt [hunk 1]
```diff
@@ -0,0 +1,13 @@
+package app.revanced.patches.youtube.layout.castbutton.annotations
+
+import app.revanced.patcher.annotation.Compatibility
+import app.revanced.patcher.annotation.Package
+
+@Compatibility(
+    [Package(
+        "com.google.android.youtube", arrayOf("17.14.35", "17.17.34", "17.19.36", "17.20.37", "17.22.36") // only 17.20.37 is tested
```
Comment (Sculas) [inline src/main/kotlin/app/revanced/patches/youtube/layout/castbutton/annotations/CastPatchCompatibility.kt:8]:
```suggestion
        "com.google.android.youtube", arrayOf("17.20.37")
```
If only `17.20.37` is tested, you should only put that in the compatibility list.
Only when a patch is tested for a certain version, may it be added to the compatibility list.

File: src/main/kotlin/app/revanced/patches/youtube/layout/castbutton/signatures/CastButtonSignature.kt [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+@Name("cast-button-signature")
+@MatchingMethod(
+    "Landroidx/mediarouter/app/MediaRouteButton", "setVisibility" // first one is prolly causing the issue since not sure what to write there
+)
+@FuzzyPatternScanMethod(2) // FIXME: Test this threshold and find the best value.
+@CastButtonCompatibility
+@Version("0.0.1")
+object CastButtonSignature : MethodSignature(
+    "V", // this might be broken as well since idk what is this
```
Comment (Sculas) [inline src/main/kotlin/app/revanced/patches/youtube/layout/castbutton/signatures/CastButtonSignature.kt:21]: This is a `void` type. See [here](https://docs.oracle.com/javase/specs/jvms/se7/html/jvms-4.html#jvms-4.3.2) for more information about types in bytecode.

File: src/main/kotlin/app/revanced/patches/youtube/layout/castbutton/signatures/CastButtonSignature.kt [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    AccessFlags.PUBLIC or AccessFlags.FINAL,
+    listOf("L","L"), // this too
```
Comment (Sculas) [inline src/main/kotlin/app/revanced/patches/youtube/layout/castbutton/signatures/CastButtonSignature.kt:23]: Like above, this is a `class` type. We do not specify what kind of class/object it is, we just say "it's a class/object".

File: src/main/kotlin/app/revanced/patches/youtube/layout/castbutton/signatures/CastButtonSignature.kt [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+import app.revanced.patcher.signature.implementation.method.MethodSignature
+import app.revanced.patcher.signature.implementation.method.annotation.FuzzyPatternScanMethod
+import app.revanced.patcher.signature.implementation.method.annotation.MatchingMethod
+import app.revanced.patches.youtube.layout.castbutton.annotations.CastButtonCompatibility
+import org.jf.dexlib2.AccessFlags
+import org.jf.dexlib2.Opcode
+
+@Name("cast-button-signature")
+@MatchingMethod(
+    "Landroidx/mediarouter/app/MediaRouteButton", "setVisibility" // first one is prolly causing the issue since not sure what to write there
```
Comment (Sculas) [inline src/main/kotlin/app/revanced/patches/youtube/layout/castbutton/signatures/CastButtonSignature.kt:15]: Why are you hooking `MediaRouteButton#setVisibility`? What is it you're trying to do with this patch?

**Review by Sculas — CHANGES_REQUESTED** [review]

File: src/main/kotlin/app/revanced/patches/youtube/layout/castbutton/signatures/CastButtonSignature.kt (hunk 3 above)
Comment (Canny1913) [reply src/main/kotlin/app/revanced/patches/youtube/layout/castbutton/signatures/CastButtonSignature.kt:23]: 👍 
are both of these correct or should i change them?

File: src/main/kotlin/app/revanced/patches/youtube/layout/castbutton/signatures/CastButtonSignature.kt (hunk 4 above)
Comment (Canny1913) [reply src/main/kotlin/app/revanced/patches/youtube/layout/castbutton/signatures/CastButtonSignature.kt:15]: ^

i basically tried to do what vanced did in their code

**Comment by Canny1913** [comment]
also i just saw that in xGlobals, castButtonShown is set to true

**Pull request merged**